README.md
# prometheus_exporters

Cookbook to install and configure various Prometheus exporters on systems to be monitored by Prometheus.

Currently supported exporters are node, postgres, redis, snmp, and wmi. More may be added in the future. Please contact the author if you have specific requests.

All of the exporters are available as chef custom resources that can be instantiated from other cookbooks.


# Supports

Ubuntu 14 & 16 (and probably other Debian based distributions)

CentOS 6 and 7 (and probably other RHEL based distributions)

Windows Server 2012 & 2016 (wmi_exporter recipe only)

# Resources

## node_exporter

* `web_listen_address`, String, default: ':9100'
* `web_telemetry_path`, String, default: '/metrics'
* `log_level`, String, default: 'info'
* `log_format`, String, default: 'logger:stdout'
* `collectors_enabled`, Array
* `collector_textfile_directory`, String
* `collector_netdev_ignored_devices`, String
* `collector_diskstats_ignored_devices`, String
* `collector_filesystem_ignored_fs_types`, String
* `collector_filesystem_ignored_mount_point`, String
* `custom_options`, String

Use `custom_options` for your configuration if defined proterties are not satisfying your needs.

```ruby

listen_ip = '127.0.0.1'

node_exporter 'main' do
  web_listen_address "#{listen_ip}:9100"
  action [:enable, :start]
end
```

or just set

* `node['prometheus_exporters']['listen_interface']`
* `node['prometheus_exporters']['node']['collectors']`
* `node['prometheus_exporters']['node']['textfile_directory']`
* `node['prometheus_exporters']['node']['ignored_net_devs']`

and add `recipe['prometheus_exporters::node]` to your run_list.

## postgres_exporter

The postgres_exporter resource supports running multiple copies of PostgreSQL exporter the same system. This is useful if you have multiple copies of PostgreSQL running on the same system
(eg. different versions) or you are connecting to multiple remote PostgreSQL servers across the network.

* `instance_name` name of PostgreSQL exporter instance. (**name attribute**)
* `data_source_name` PostgreSQL connection string. E.g. `postgresql://login:password@hostname:port/dbname`
* `extend_query_path` Path to custom queries to run
* `log_format` If set use a syslog logger or JSON logging. Example: logger:syslog?appname=bob&local=7 or logger:stdout?json=true. Defaults to stderr.
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal].
* `web_listen_address` Address to listen on for web interface and telemetry. (default "127.0.0.1:9187")
* `web_telemetry_path` Path under which to expose metrics. (default "/metrics")
* `user` System user to run exporter as. (default "postgres")

```ruby
postgres_exporter '9.5_main' do
  data_source_name 'postgresql://localhost:5432/example'
  user 'postgres'
end
```

## mysqld_exporter

The mysqld_exporter resource supports running multiple copies of the MySQL exporter on the same system.

* `instance_name` name of MySQL exporter instance. (**name attribute**)
* `data_source_name` MySQL connection string
* `config_my_cnf` Path to .my.cnf file to read MySQL credentials from. (default: ~/.my.cnf)
* `log_format` If set use a syslog logger or JSON logging. Example: logger:syslog?appname=bob&local=7 or logger:stdout?json=true. Defaults to stderr.
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal].
* `web_listen_address` Address to listen on for web interface and telemetry. (default "127.0.0.1:9104")
* `web_telemetry_path` Path under which to expose metrics. (default "/metrics")
* `user` System user to run exporter as. (default "mysql")

```ruby
mysqld_exporter 'main' do
  data_source_name '/'
  config_my_cnf '~/.my/cnf'
  user 'mysql'
end
```

## wmi_exporter

Expects the Chocolatey package manager to already be installed.  This is up to individuals to provide by including the [Chocolatey cookbook](https://github.com/chocolatey/chocolatey-cookbook) in their own wrapper cookbooks.

* `version`, String, default: '0.2.7'
* `enabled_collectors`, String, default: 'cpu,cs,logical_disk,net,os,service,system'
* `listen_address`, String, default: '0.0.0.0'
* `listen_port`, String, default: '9182'
* `metrics_path`, Strin, default: '/metrics'

Use the given defaults or set the attributes...

* `node['prometheus_exporters']['wmi']['version']['listen_interface']`
* `node['prometheus_exporters']['wmi']['listen_address']`
* `node['prometheus_exporters']['wmi']['listen_port']`
* `node['prometheus_exporters']['wmi']['metrics_path']`

and add `recipe['prometheus_exporters::wmi]` to your run_list.

# Discovery

Each exporter will set an attribute when it's enabled, in the form of `node['prometheus_exporters'][exporter_name]['enabled']`. This makes it possible to search for
exporters within your environment using knife search or from within other cookbooks using a query such as:

```knife search node 'prometheus_exporters_node_enabled:true'```

This query will return all nodes with configured node exporters which can be used for automatically configuring Prometheus servers.

# Known Issues

* The snmp_exporter requires a configuration file that is usually created by a config generator. Currently this functionality must be provided by a wrapper cookbook.


CHANGELOG.md
# prometheus_exporters

## 0.5.2
- [Viktor Radnai] - Bugfix: race condition with redis_exporter binary missing before systemd trying to verify it
- [Viktor Radnai] - Fix deprecation warnings
- [Viktor Radnai] - Bugfix: quote environment variables in sysV init script

## 0.5.1
- [Kieren Scott] - MySQL database exporter resource

## 0.5.0
- [Matt Mencel] - wmi_exporter recipe
- [Viktor Radnai] - Adding 'enabled' node property for each exporter for use with Chef search

## 0.4.2
- [Viktor Radnai] - Bugfix for starting redis process as the specified user

## 0.4.1

- [Viktor Radnai] - Split log directories to fix logfile permission issue when exporters aren't running as root
- [Viktor Radnai] - Bugfixes for postgres_exporter
- [Viktor Radnai] - Improvements for tests
- [Viktor Radnai] - Added checksum for SNMP exporter

## 0.4.0

- [Viktor Radnai] - Added tests
- [Viktor Radnai] - Improved service configuration for all exporters
- [Viktor Radnai] - Fixed errors reported by `cookstyle`
- [Viktor Radnai] - Merged changes from Matt Mencel's repo (https://github.com/WIU/chef-prometheus-exporters.git)
- [Matt Mencel] - CentOS Support: updated only_if and not_if in node_exporter service
- [Matt Mencel] - Chef Warnings: Fixed Chef warnings in upstart template

## 0.3.0

- [Kirill Kuznetsov] - node_exporter version bump: `0.15.0`
- [Kirill Kuznetsov] - BREAKING: the new cmd syntax brings the need for a new resource properties.

## 0.2.1

- [Kirill Kuznetsov] - enable weave network interface monitoring by default to alert on its stauts.
- [Kirill Kuznetsov] - it now depends on `systemd` cookbook in general

## 0.2.0

- [Kirill Kuznetsov] - new systemd cookbook with new lwrp syntax
- [Kirill Kuznetsov] - new `redis_exporter` resource to install redis_exporter.
- [Kirill Kuznetsov] - set correct mode for postgres_exporter executable
- [Kirill Kuznetsov] - surround attributes for postgresql exporter in quotes.

## 0.1.3

- [Matt Mencel] - CentOS Support: updated only_if and not_if in node_exporter service
- [Matt Mencel] - Chef Warnings: Fixed Chef warnings in upstart template

## 0.1.2

- [Kirill Kuznetsov] - PostgreSQL Exporter resource.
- [Kirill Kuznetsov] - more options for upstart service template: env & setuid.
- [Kirill Kuznetsov] - default ignored mount points for Node Exporter.

## 0.1.0
- [Kirill Kuznetsov] - Initial release with `node_exporter` support.


test/integration/default/default_test.rb
# Node exporter
describe port(9100) do
  it { should be_listening }
  its('processes') { should cmp 'node_exporter' }
end

describe service('node_exporter') do
  it { should be_enabled }
  it { should be_running }
end

# Mysqld exporter
describe port(9104) do
  it { should be_listening }
  its('processes') { should cmp(/^mysqld_expo/) }
end

describe service('mysqld_exporter_main') do
  it { should be_enabled }
  it { should be_running }
end

# Redis exporter
describe port(9121) do
  it { should be_listening }
  its('processes') { should cmp(/^redis_expo/) }
end

describe service('redis_exporter') do
  it { should be_enabled }
  it { should be_running }
end

# Postgres exporter
describe port(9187) do
  it { should be_listening }
  its('processes') { should cmp(/^postgres_expo/) }
end

describe service('postgres_exporter_main') do
  it { should be_enabled }
  it { should be_running }
end

describe port(9188) do
  it { should be_listening }
  its('processes') { should cmp(/^postgres_expo/) }
end

describe service('postgres_exporter_chef') do
  it { should be_enabled }
  it { should be_running }
end

describe processes('postgres_exporter') do
  its('users') { should eq ['postgres', 'opscode-pgsql'] }
end

# SNMP exporter
describe service('postgres_exporter_chef') do
  it { should be_enabled }
end


test/integration/cookbooks/testrig/recipes/default.rb
directory '/etc/mysqld' do
  action :create
end

file '/etc/mysqld/my.cnf' do
  content 'user		= mysql
pid-file	= /var/run/mysqld/mysqld.pid
socket		= /var/run/mysqld/mysqld.sock
port		= 3306
basedir		= /usr
datadir		= /var/lib/mysql
'
  action :create
end
user 'mysql' do
  comment 'Mock Mysql user'
  system true
end
user 'prometheus' do
  comment 'Prometheus user'
  system true
end
user 'postgres' do
  comment 'Mock PostgreSQL user'
  system true
end
user 'opscode-pgsql' do
  comment 'Mock Chef PostgreSQL user'
  system true
end

mysqld_exporter 'main' do
  data_source_name '/'
  config_my_cnf '/etc/mysqld/my.cnf'
  user 'mysql'

  action [:install, :enable, :start]
end

node_exporter 'main' do
  collectors_enabled node['prometheus_exporters']['node']['collectors_enabled']
  collectors_disabled node['prometheus_exporters']['node']['collectors_disabled']
  collector_textfile_directory node['prometheus_exporters']['node']['textfile_directory']
  collector_netdev_ignored_devices node['prometheus_exporters']['node']['ignored_net_devs']
  collector_filesystem_ignored_mount_points node['prometheus_exporters']['node']['ignored_mount_points']

  action [:install, :enable, :start]
end

redis_exporter 'main' do
  action [:install, :enable, :start]
end

snmp_exporter 'main' do
  action [:install, :enable] # Will not start without a valid snmp.yml config file
end

postgres_exporter 'main' do
  data_source_name 'postgres'
  action [:install, :enable, :start]
end

# Second PostgreSQL exporter instance with the settings needed to
# monitor the PostgreSQL server bundled with Chef server
postgres_exporter 'chef' do
  web_listen_address '0.0.0.0:9188'
  data_source_name 'user=opscode-pgsql host=/tmp/ sslmode=disable'
  user 'opscode-pgsql'

  action [:install, :enable, :start]
end


resources/redis.rb
#
# Cookbook Name:: prometheus_exporters
# Resource:: redis
#
# Copyright 2017, Evil Martians
#
# All rights reserved - Do Not Redistribute
#

resource_name :redis_exporter

property :web_listen_address, String, default: '0.0.0.0:9121'
property :web_telemetry_path, String, default: '/metrics'
property :log_format, String, default: 'txt'
property :debug, [TrueClass, FalseClass], default: false
property :check_keys, String
property :redis_addr, String, default: 'redis://localhost:6379'
property :redis_password, String
property :redis_alias, String
property :redis_file, String
property :namespace, String, default: 'redis'
property :user, String, default: 'root'

action :install do
  # Set property that can be queried with Chef search
  node.default['prometheus_exporters']['redis']['enabled'] = true

  options = "-web.listen-address #{new_resource.web_listen_address}"
  options += " -web.telemetry-path #{new_resource.web_telemetry_path}"
  options += " -log-format #{new_resource.log_format}"
  options += ' -debug' if new_resource.debug
  options += " -check-keys #{new_resource.check_keys}" if new_resource.check_keys
  options += " -redis.addr #{new_resource.redis_addr}" if new_resource.redis_addr
  options += " -redis.password #{new_resource.redis_password}" if new_resource.redis_password
  options += " -redis.alias #{new_resource.redis_alias}" if new_resource.redis_alias
  options += " -redis.file #{new_resource.redis_file}" if new_resource.redis_file
  options += " -namespace #{new_resource.namespace}"

  remote_file 'redis_exporter' do
    path "#{Chef::Config[:file_cache_path]}/redis_exporter.tar.gz"
    owner 'root'
    group 'root'
    mode '0644'
    source node['prometheus_exporters']['redis']['url']
    checksum node['prometheus_exporters']['redis']['checksum']
  end

  bash 'untar redis_exporter' do
    code "tar -xzf #{Chef::Config[:file_cache_path]}/redis_exporter.tar.gz -C /usr/local/sbin/"
    action :nothing
    subscribes :run, 'remote_file[redis_exporter]', :immediately
  end

  service 'redis_exporter' do
    action :nothing
  end

  case node['init_package']
  when /init/
    %w(
      /var/run/prometheus
      /var/log/prometheus
    ).each do |dir|
      directory dir do
        owner 'root'
        group 'root'
        mode '0755'
        recursive true
        action :create
      end
    end

    directory '/var/log/prometheus/redis_exporter' do
      owner new_resource.user
      group 'root'
      mode '0755'
      action :create
    end

    template '/etc/init.d/redis_exporter' do
      cookbook 'prometheus_exporters'
      source 'initscript.erb'
      owner 'root'
      group 'root'
      mode '0755'
      variables(
        name: 'redis_exporter',
        user: new_resource.user,
        cmd: "/usr/local/sbin/redis_exporter #{options}",
        service_description: 'Prometheus Redis Exporter'
      )
      notifies :restart, 'service[redis_exporter]'
    end

  when /systemd/
    systemd_unit 'redis_exporter.service' do
      content(
        'Unit' => {
          'Description' => 'Systemd unit for Prometheus Redis Exporter',
          'After' => 'network.target remote-fs.target apiserver.service',
        },
        'Service' => {
          'Type' => 'simple',
          'User' => new_resource.user,
          'ExecStart' => "/usr/local/sbin/redis_exporter #{options}",
          'WorkingDirectory' => '/',
          'Restart' => 'on-failure',
          'RestartSec' => '30s',
        },
        'Install' => {
          'WantedBy' => 'multi-user.target',
        }
      )
      notifies :restart, 'service[redis_exporter]'
      action :create
    end

  when /upstart/
    template '/etc/init/redis_exporter.conf' do
      cookbook 'prometheus_exporters'
      source 'upstart.conf.erb'
      owner 'root'
      group 'root'
      mode '0644'
      variables(
        env: environment_list,
        user: new_resource.user,
        cmd: "/usr/local/sbin/redis_exporter #{options}",
        service_description: 'Prometheus Redis Exporter'
      )
      notifies :restart, 'service[redis_exporter]'
    end

  else
    raise "Init system '#{node['init_package']}' is not supported by the 'prometheus_exporters' cookbook"
  end
end

action :enable do
  action_install
  service 'redis_exporter' do
    action :enable
  end
end

action :start do
  service 'redis_exporter' do
    action :start
  end
end

action :disable do
  service 'redis_exporter' do
    action :disable
  end
end

action :stop do
  service 'redis_exporter' do
    action :stop
  end
end


resources/postgres.rb
#
# Cookbook Name:: prometheus_exporters
# Resource:: postgres
#
# Copyright 2017, Evil Martians
#
# All rights reserved - Do Not Redistribute
#

resource_name :postgres_exporter

property :instance_name, String, name_property: true
property :data_source_name, String, required: true
property :extend_query_path, String
property :log_format, String, default: 'logger:stdout?json=false'
property :log_level, String
property :web_listen_address, String, default: '0.0.0.0:9187'
property :web_telemetry_path, String
property :user, String, default: 'postgres'

action :install do
  # Set property that can be queried with Chef search
  node.default['prometheus_exporters']['postgres']['enabled'] = true

  service_name = "postgres_exporter_#{new_resource.instance_name}"

  options = "-web.listen-address '#{new_resource.web_listen_address}'"
  options += " -web.telemetry-path '#{new_resource.web_telemetry_path}'" if new_resource.web_telemetry_path
  options += " -log.level #{new_resource.log_level}" if new_resource.log_level
  options += " -log.format '#{new_resource.log_format}'"
  options += " -extend.query-path #{new_resource.extend_query_path}" if new_resource.extend_query_path

  env = {
    'DATA_SOURCE_NAME' => new_resource.data_source_name,
  }

  remote_file 'postgres_exporter' do
    path '/usr/local/sbin/postgres_exporter'
    owner 'root'
    group 'root'
    mode '0755'
    source node['prometheus_exporters']['postgres']['url']
    checksum node['prometheus_exporters']['postgres']['checksum']
  end

  service service_name do
    action :nothing
  end

  case node['init_package']
  when /init/
    %w(
      /var/run/prometheus
      /var/log/prometheus
    ).each do |dir|
      directory dir do
        owner 'root'
        group 'root'
        mode '0755'
        recursive true
        action :create
      end
    end

    directory "/var/log/prometheus/#{service_name}" do
      owner new_resource.user
      group 'root'
      mode '0755'
      action :create
    end

    template "/etc/init.d/#{service_name}" do
      cookbook 'prometheus_exporters'
      source 'initscript.erb'
      owner 'root'
      group 'root'
      mode '0755'
      variables(
        env: env,
        user: new_resource.user,
        name: service_name,
        cmd: "/usr/local/sbin/postgres_exporter #{options}",
        service_description: 'Prometheus PostgreSQL Exporter'
      )
      notifies :restart, "service[#{service_name}]"
    end

  when /systemd/
    systemd_unit "#{service_name}.service" do
      content(
        'Unit' => {
          'Description' => 'Systemd unit for Prometheus PostgreSQL Exporter',
          'After' => 'network.target remote-fs.target apiserver.service',
        },
        'Service' => {
          'Type' => 'simple',
          'User' => new_resource.user,
          'ExecStart' => "/usr/local/sbin/postgres_exporter #{options}",
          'Environment' => env.map { |k, v| "'#{k}=#{v}'" }.join(' '),
          'WorkingDirectory' => '/',
          'Restart' => 'on-failure',
          'RestartSec' => '30s',
        },
        'Install' => {
          'WantedBy' => 'multi-user.target',
        }
      )
      notifies :restart, "service[#{service_name}]"
      action :create
    end

  when /upstart/
    template "/etc/init/#{service_name}.conf" do
      cookbook 'prometheus_exporters'
      source 'upstart.conf.erb'
      owner 'root'
      group 'root'
      mode '0644'
      variables(
        env: env,
        setuid: new_resource.user,
        cmd: "/usr/local/sbin/postgres_exporter #{options}",
        service_description: 'Prometheus PostgreSQL Exporter'
      )
      notifies :restart, "service[#{service_name}]"
    end

  else
    raise "Init system '#{node['init_package']}' is not supported by the 'prometheus_exporters' cookbook"
  end
end

action :enable do
  action_install
  service "postgres_exporter_#{new_resource.instance_name}" do
    action :enable
  end
end

action :start do
  service "postgres_exporter_#{new_resource.instance_name}" do
    action :start
  end
end

action :disable do
  service "postgres_exporter_#{new_resource.instance_name}" do
    action :disable
  end
end

action :stop do
  service "postgres_exporter_#{new_resource.instance_name}" do
    action :stop
  end
end


resources/mysqld.rb
#
# Cookbook Name:: prometheus_exporters
# Resource:: mysqld
#
# Copyright 2017, Evil Martians
#
# All rights reserved - Do Not Redistribute
#

resource_name :mysqld_exporter

property :instance_name, String, name_property: true
property :data_source_name, String, required: true
property :log_format, String, default: 'logger:stdout?json=false'
property :log_level, String
property :web_listen_address, String, default: '0.0.0.0:9104'
property :web_telemetry_path, String
property :config_my_cnf, String, default: '~/.my.cnf'
property :user, String, default: 'mysql'

action :install do
  # Set property that can be queried with Chef search
  node.default['prometheus_exporters']['mysqld']['enabled'] = true

  service_name = "mysqld_exporter_#{instance_name}"

  options = "-web.listen-address '#{web_listen_address}'"
  options += " -web.telemetry-path '#{web_telemetry_path}'" if web_telemetry_path
  options += " -config.my-cnf '#{config_my_cnf}'" if config_my_cnf
  options += " -log.level #{log_level}" if log_level
  options += " -log.format '#{log_format}'"

  env = {
    'DATA_SOURCE_NAME' => data_source_name,
  }

  remote_file 'mysqld_exporter' do
    path "#{Chef::Config[:file_cache_path]}/mysqld_exporter.tar.gz"
    owner 'root'
    group 'root'
    mode '0644'
    source node['prometheus_exporters']['mysqld']['url']
    checksum node['prometheus_exporters']['mysqld']['checksum']
  end

  bash 'untar mysqld_exporter' do
    code "tar -xzf #{Chef::Config[:file_cache_path]}/mysqld_exporter.tar.gz -C /opt"
    action :nothing
    subscribes :run, 'remote_file[mysqld_exporter]', :immediately
  end

  link '/usr/local/sbin/mysqld_exporter' do
    to "/opt/mysqld_exporter-#{node['prometheus_exporters']['mysqld']['version']}.linux-amd64/mysqld_exporter"
  end

  service service_name do
    action :nothing
  end

  case node['init_package']
  when /init/
    %w(
      /var/run/prometheus
      /var/log/prometheus
    ).each do |dir|
      directory dir do
        owner 'root'
        group 'root'
        mode '0755'
        recursive true
        action :create
      end
    end

    directory "/var/log/prometheus/#{service_name}" do
      owner new_resource.user
      group 'root'
      mode '0755'
      action :create
    end

    template "/etc/init.d/#{service_name}" do
      cookbook 'prometheus_exporters'
      source 'initscript.erb'
      owner 'root'
      group 'root'
      mode '0755'
      variables(
        env: env,
        user: new_resource.user,
        name: service_name,
        cmd: "/usr/local/sbin/mysqld_exporter #{options}",
        service_description: 'Prometheus MySQL Exporter'
      )
      notifies :restart, "service[#{service_name}]"
    end

  when /systemd/
    systemd_unit "#{service_name}.service" do
      content(
        'Unit' => {
          'Description' => 'Systemd unit for Prometheus MySQL Exporter',
          'After' => 'network.target remote-fs.target apiserver.service',
        },
        'Service' => {
          'Type' => 'simple',
          'User' => new_resource.user,
          'ExecStart' => "/usr/local/sbin/mysqld_exporter #{options}",
          'Environment' => env.map { |k, v| "'#{k}=#{v}'" }.join(' '),
          'WorkingDirectory' => '/',
          'Restart' => 'on-failure',
          'RestartSec' => '30s',
        },
        'Install' => {
          'WantedBy' => 'multi-user.target',
        }
      )
      notifies :restart, "service[#{service_name}]"
      action :create
    end

  when /upstart/
    template "/etc/init/#{service_name}.conf" do
      cookbook 'prometheus_exporters'
      source 'upstart.conf.erb'
      owner 'root'
      group 'root'
      mode '0644'
      variables(
        env: env,
        setuid: new_resource.user,
        cmd: "/usr/local/sbin/mysqld_exporter #{options}",
        service_description: 'Prometheus MySQL Exporter'
      )
      notifies :restart, "service[#{service_name}]"
    end

  else
    raise "Init system '#{node['init_package']}' is not supported by the 'prometheus_exporters' cookbook"
  end
end

action :enable do
  action_install
  service "mysqld_exporter_#{instance_name}" do
    action :enable
  end
end

action :start do
  service "mysqld_exporter_#{instance_name}" do
    action :start
  end
end

action :disable do
  service "mysqld_exporter_#{instance_name}" do
    action :disable
  end
end

action :stop do
  service "mysqld_exporter_#{instance_name}" do
    action :stop
  end
end
